packages/builder/src/stores/portal/groups.ts

```diff
@@ -0,0 +1,93 @@
+import { get } from "svelte/store"
+import { API } from "@/api"
+import { licensing } from "@/stores/portal"
+import { UserGroup } from "@budibase/types"
+import { BudiStore } from "../BudiStore"
+
+class GroupStore extends BudiStore<UserGroup[]> {
+  constructor() {
+    super([])
+  }
+
+  updateStore = (group: UserGroup) => {
+    this.update(state => {
+      const currentIdx = state.findIndex(gr => gr._id === group._id)
+      if (currentIdx >= 0) {
+        state.splice(currentIdx, 1, group)
+      } else {
+        state.push(group)
+      }
+      return state
+    })
+  }
+
+  async init() {
+    // Only init if there is a groups license, just to be sure but the feature will be blocked
+    // on the backend anyway
+    if (get(licensing).groupsEnabled) {
+      const groups = await API.getGroups()
+      this.set(groups)
+    }
+  }
+
+  private async refreshGroup(groupId: string) {
+    const group = await API.getGroup(groupId)
+    this.updateStore(group)
+  }
+
+  async save(group: UserGroup) {
+    const { ...dataToSave } = group
+    delete dataToSave.scimInfo
+    const response = await API.saveGroup(dataToSave)
+    group._id = response._id
+    group._rev = response._rev
+    this.updateStore(group)
+    return group
+  }
+
+  async delete(group: UserGroup) {
+    await API.deleteGroup(group._id!, group._rev!)
+    this.update(state => {
+      state = state.filter(state => state._id !== group._id)
```

Comment:
More of an academic question than something to act on: does this copy `state`? Is there a way to do it without copying? e.g. `state.find` followed by `state.splice`?

Something I was extremely surprised by in the course of writing my queueing blog post is just how fast `array.splice` is in JS. It appears to be O(1), iirc.

Reply:
https://jsbm.dev/9sJxjgmMBsVe5

Reply:
That's a very nice benchmarking site - TIL! Interesting that find/splice is about 2x faster. It actually got slightly  faster again (15 op/s to 16 on this M1 MBP) when I replaced `array.find` with `array.findIndex`, which I think is what you meant to use.